### PR TITLE
Debug: Improve RunningTask.__repr__

### PR DESCRIPTION
**Pull request by garmin-mjames**
Add `RunningTask.__repr__` that shows current status and unique task id.

This is the base for a follow-on PR to overhaul scheduler debug.

Each `RunningTask` has it's own "coroutine stack", and currently that's not easily accessible, so it can be difficult during debugging to figure out the state and call location of a task. Generator coroutines can `yield from` other generators and coroutines can `await` other coroutines, both of which will bypass the scheduler.

Depending on the status of the task, the pending trigger object or outcome will also be displayed.

Example `repr()` outputs from `test_task_repr`:

- `<Task 2 created coro=generator_coro_outer()>`
- `<Task 2 pending coro=generator_coro_inner() trigger=<Timer of 1000.00ps at 0x0F0B2A58>>`
- `<Task 2 finished coro=generator_coro_outer() outcome=Error(ValueError('inner'))>`
- `<Task 3 adding coro=coroutine_outer()>`
- `<Task 3 running coro=coroutine_outer()>`
- ~~`<Task 3 pending coro=coroutine_inner() trigger=<<Event for Combine(<Task 4 pending coro=coroutine_wait() trigger=<Timer of 1000.00ps at 0x0F418210>>, <Task 5 pending coro=coroutine_wait() trigger=<Timer of 1000.00ps at 0x0F0B2AA8>>) at 0x0F0A038B>.wait() at 0x0F0B29B8>>`~~
- `<Task 3 pending coro=coroutine_inner() trigger=Combine(Join(<Task 4>), Join(<Task 5>)>`

Tests are displayed `<Test test_name ...>` rather than with the task id.

The current scheduler debug output is less useful after this PR, as `__name__` and `__qualname__` are repurposed, so I changed it to get them from the coroutine wrapped by `RunningTask`.

### CODE REVIEW

File: cocotb/utils.py [hunk 1]
```diff
@@ -578,3 +579,36 @@ def remove_traceback_frames(tb_or_exc, frame_names):
...
+                f = getattr(coro, 'gi_frame')
+                coro = coro.gi_yieldfrom
+            except AttributeError:
+                f = None
+                coro = None
+        if f is not None:
+            yield (f, f.f_lineno)
+
+
+def extract_coro_stack(coro, limit=None):
```
Comment (garmin-mjames) [inline cocotb/utils.py:601]: This is based on `traceback.extract_tb`, using the `StackSummary.extract` classmethod for ease of use.
I think it might be possible to get the `__qualname__` for each coroutine in the stack and display that.

File: cocotb/decorators.py [hunk 2]
```diff
@@ -128,7 +134,50 @@ def __iter__(self):
         return self
 
     def __str__(self):
-        return str(self.__qualname__)
+        return "<{}>".format(self.__name__)
+
+    def _get_coro_stack(self):
+        """
+        Get the coroutine callstack of this Task.
+        """
```
Comment (cmarqu) [inline cocotb/decorators.py:142]:
```suggestion
        """Get the coroutine callstack of this Task."""
```
https://www.python.org/dev/peps/pep-0257/#one-line-docstrings (because I don't have to say anything substantial :)

File: cocotb/scheduler.py [hunk 3]
```diff
@@ -81,6 +83,15 @@ def __exit__(self, *excinfo):
         _profile.disable()
 
 
+@contextmanager
+def task_context(scheduler, task):
+    """Context manager for currently running task."""
```
Comment (cmarqu) [inline cocotb/scheduler.py:88]:
```suggestion
    """Context manager for the currently running task."""
```

File: cocotb/utils.py [hunk 4]
```diff
@@ -578,3 +579,36 @@ def remove_traceback_frames(tb_or_exc, frame_names):
             assert tb.tb_frame.f_code.co_name == frame_name
             tb = tb.tb_next
         return tb
+
+
+def walk_coro_stack(coro):
+    """Walk down the coroutine stack. Supports coroutines and generators."""
```
Comment (cmarqu) [inline cocotb/utils.py:585]:
```suggestion
    """Walk down the coroutine stack, starting at *coro*.
    
    Supports coroutines and generators."""
```
?

File: cocotb/utils.py [hunk 5]
```diff
@@ -578,3 +579,36 @@ def remove_traceback_frames(tb_or_exc, frame_names):
...
+            yield (f, f.f_lineno)
+
+
+def extract_coro_stack(coro, limit=None):
+    """
+    Return a :class:`traceback.StackSummary` object representing a list of
+    pre-processed entries from traceback.
+
+    This is useful for alternate formatting of stack traces.  If
+    'limit' is omitted or None, all entries are extracted.  A
```
Comment (cmarqu) [inline cocotb/utils.py:607]:
```suggestion
    *limit* is omitted or None, all entries are extracted.  A
```

File: cocotb/utils.py [hunk 6]
```diff
@@ -578,3 +579,36 @@ def remove_traceback_frames(tb_or_exc, frame_names):
...
+                coro = coro.gi_yieldfrom
+            except AttributeError:
+                f = None
+                coro = None
+        if f is not None:
+            yield (f, f.f_lineno)
+
+
+def extract_coro_stack(coro, limit=None):
+    """
```
Comment (cmarqu) [inline cocotb/utils.py:602]: There should be one-line summary on the same line as the opening """.

File: cocotb/utils.py [hunk 7]
```diff
@@ -578,3 +579,36 @@ def remove_traceback_frames(tb_or_exc, frame_names):
...
+    Return a :class:`traceback.StackSummary` object representing a list of
+    pre-processed entries from traceback.
+
+    This is useful for alternate formatting of stack traces.  If
+    'limit' is omitted or None, all entries are extracted.  A
+    pre-processed stack trace entry is a FrameSummary object
+    containing attributes filename, lineno, name, and line
```
Comment (cmarqu) [inline cocotb/utils.py:609]:
```suggestion
    pre-processed stack trace entry is a :class:`traceback.FrameSummary` object
    containing the attributes ``filename``, ``lineno``, ``name``, and ``line``
```

File: tests/test_cases/test_cocotb/test_scheduler.py [hunk 8]
```diff
@@ -263,3 +264,81 @@ async def waiter(event):
     await NullTrigger()
 
     assert waiter_scheduled is True
+
+
+@cocotb.test()
+async def test_task_repr(dut):
+    """
+    Test RunningTask.__repr__
+    """
```
Comment (cmarqu) [inline tests/test_cases/test_cocotb/test_scheduler.py:273]:
```suggestion
    """Test RunningTask.__repr__"""
```
(Our tests are not usually that polished, but why not start :)

File: cocotb/decorators.py [hunk 9]
```diff
@@ -128,7 +134,50 @@ def __iter__(self):
...
+        if cocotb.scheduler._current_task is self:
+            fmt = "<{name} running coro={coro}()>"
+        elif self._finished:
+            fmt = "<{name} finished coro={coro}() outcome={outcome}>"
+        elif self._trigger is not None:
+            fmt = "<{name} pending coro={coro}() trigger={trigger}>"
+        elif not self._started:
+            fmt = "<{name} created coro={coro}()>"
+        else:
+            fmt = "<{name} adding coro={coro}()>"
```
Comment (garmin-mjames) [inline cocotb/decorators.py:164]: While the coroutine is running, we can't get the full stack, so for `running` and `adding` states the initial coroutine name is printed.

File: tests/test_cases/test_cocotb/test_scheduler.py [hunk 10]
```diff
@@ -300,3 +301,78 @@ async def second():
     await ReadOnly()
 
     assert dut.array_7_downto_4.value == [10, 2, 3, 4]
+
+
+async def test_task_repr(dut):
```
Comment (ktbarrett) [inline tests/test_cases/test_cocotb/test_scheduler.py:306]:
```suggestion
@cocotb.test()
async def test_task_repr(dut):
```

File: cocotb/triggers.py [hunk 11]
```diff
@@ -696,7 +696,7 @@ class Combine(_AggregateWaitable):
 
     async def _wait(self):
         waiters = []
-        e = Event()
+        e = Event(self)
```
Comment (ktbarrett) [inline cocotb/triggers.py:699]: I don't think the abuse is a big issue, this object cannot directly leak to the user, but I think this transition could be smoother so we don't see the `Event` at all, it is an implementation detail. I think that would require a special subclass of `Event`. Sounds like a follow up PR.

File: tests/test_cases/test_cocotb/test_scheduler.py (hunk 10 above)
Comment (garmin-mjames) [reply tests/test_cases/test_cocotb/test_scheduler.py:306]: whoops, missed that in the rebase.

**Comment by garmin-mjames** [comment]
@ktbarrett 
> I can imagine each stack level could be a newline and indentation, and trees in the execution due to `Combine` or `First` would be on new lines without indentation.

Where are you imagining this would be printed?

**Comment by ktbarrett** [comment]
That's up to the user, no? Presumably the log. Maybe I'm not understanding what you're saying.

**Comment by garmin-mjames** [comment]
The stack returned from `_get_coro_stack` is only the "task-local" Python coroutine stack outside of the scheduler. Anything involving triggers has to involve the `scheduler`, as that's cross-task. This PR gets us the actual coroutine that has awaited a `Trigger`, not just the name of the coroutine that the task was started with.

I have another follow-on PR for the scheduler debug output that shows the different tasks and what triggers they are pending on when cocotb returns to the simulator.

**Comment by ktbarrett** [comment]
I was just saying I would prefer better formatting.

File: cocotb/decorators.py [hunk 12]
```diff
@@ -237,19 +284,24 @@ def handle(self, record):
...
         self.start_time = 0
         self.start_sim_time = 0
         self.expect_fail = parent.expect_fail
         self.expect_error = parent.expect_error
         self.skip = parent.skip
         self.stage = parent.stage
-        self._id = parent._id
+        self._test_id = parent._id
+        self.__name__ = "Test %s" % inst.__name__
+        self.__qualname__ = "Test %s" % inst.__qualname__
```
[the remark this answers is not in the document]
Comment (garmin-mjames) [reply cocotb/decorators.py:297]: You could overwrite them in `RegressionManager._init_test` before starting the test.

File: cocotb/scheduler.py [hunk 13]
```diff
@@ -81,6 +83,15 @@ def __exit__(self, *excinfo):
         _profile.disable()
 
 
+@contextmanager
+def task_context(scheduler, task):
```
Comment (ktbarrett) [inline cocotb/scheduler.py:87]: Is there a particular reason this is a free function and not a method on `Scheduler`?

File: cocotb/decorators.py [hunk 14]
```diff
@@ -238,19 +284,24 @@ def handle(self, record):
...
+        self.log = SimLog("cocotb.test.%s" % inst.__qualname__, id(self))
         self.started = False
         self.start_time = 0
         self.start_sim_time = 0
         self.expect_fail = parent.expect_fail
         self.expect_error = parent.expect_error
         self.skip = parent.skip
         self.stage = parent.stage
-        self._id = parent._id
+        self._test_id = parent._id
```
Comment (ktbarrett) [inline cocotb/decorators.py:295]: What's the reason for this change? I don't see it anywhere else.

Comment (garmin-mjames) [reply cocotb/decorators.py:295]: It's an artifact from when all tests were initialized during test discovery, leftover after rebase.

I wanted to sort based on the id's of the `test` object not the id of the created `RunningTask`, so each `RunningTest` would have an `_id` and a `_test_id`, and `RegressionManager` would sort via the latter. I'll remove it as it's no longer needed.

File: cocotb/scheduler.py (hunk 13 above)
Comment (garmin-mjames) [reply cocotb/scheduler.py:87]: I originally tried a custom context manager like `profiling_context` above, but couldn't get it to work correctly, then I found `@contextmanager`. I haven't tried as a method on `Scheduler`.

Comment (eric-wieser) [reply cocotb/scheduler.py:87]: `@contextmanager` works just fine on methods

File: cocotb/scheduler.py [hunk 15]
```diff
@@ -81,6 +83,15 @@ def __exit__(self, *excinfo):
...
+    """Context manager for the currently running task."""
+    old_task = scheduler._current_task
+    scheduler._current_task = task
+    yield
+    scheduler._current_task = old_task
```
Comment (eric-wieser) [inline cocotb/scheduler.py:92]:
```suggestion
    scheduler._current_task = task
    try:
        yield
    finally:
        scheduler._current_task = old_task
```

File: cocotb/scheduler.py [hunk 16]
```diff
@@ -246,6 +254,9 @@ def __init__(self):
         self._test = None
         self._main_thread = threading.current_thread()
 
+        self._current_task = None
+        self._task_context = partial(task_context, self)
```
Comment (eric-wieser) [inline cocotb/scheduler.py:258]:
```suggestion
```
once you move `task_context` to a private method

**Comment by garmin-mjames** [comment]
I'm not really satisfied with the implementation of `_InternalEvent`. It should only be used by cocotb internals, but it doesn't properly enforce the single-coroutine waiting because a coroutine could `await e` instead of `await e.wait()` and it would still work.

**Comment by ktbarrett** [comment]
@garmin-mjames You are worried about the inherited implementation of `__await__` in `PythonTrigger` not doing the single-user enforcement check? You could implement `__await__` as a proxy to `.wait()`?

**Comment by garmin-mjames** [comment]
@ktbarrett I'll look at that, thanks.

File: cocotb/decorators.py [hunk 17]
```diff
@@ -87,6 +87,8 @@ class RunningTask:
         triggers to fire.
     """
 
+    _id_count = 0  # used by the scheduler for debug
```
Comment (ktbarrett) [inline cocotb/decorators.py:90]: What are the merits of this over `id`?

File: cocotb/triggers.py [hunk 18]
```diff
@@ -408,7 +408,7 @@ def wait(self):
         """
         if self.fired:
             return NullTrigger(name="{}.wait()".format(str(self)))
-        return _Event(self)
+        return _EventWaiter(self)
```
Comment (ktbarrett) [inline cocotb/triggers.py:411]: This probably doesn't need to come in.

File: cocotb/decorators.py (hunk 17 above)
Comment (garmin-mjames) [reply cocotb/decorators.py:90]: I think the main one is this from the docs:
> Two objects with non-overlapping lifetimes may have the same `id()` value.
But aesthetically I prefer a smaller integer and I think it's easier to match up visually when looking at log outputs.

**Comment by ktbarrett** [comment]
@garmin-mjames One last thing. Could you add a test/example of `Combine` to the test_task_repr?

**Comment by garmin-mjames** [comment]
> 
> 
> @garmin-mjames One last thing. Could you add a test/example of `Combine` to the test_task_repr?

There's a `Combine` in `test_task_repr`. Did you mean `First`?

**Comment by ktbarrett** [comment]
@garmin-mjames nvm me, maybe I typo'd the Ctrl+f. All my reviews seem to end the same way. Losing my mind...

**Comment by garmin-mjames** [comment]
@ktbarrett All good. I added `First` for good measure.

**Comment by garmin-mjames** [comment]
Added another case to `test_task_repr` to get full coverage of `_get_coro_stack`.

**Comment by garmin-mjames** [comment]
What about this case:

Both `<Task 1>` and `<Task 2>` wait on the same `Combine(Timer(1), Join(<Task 3>)` object. They will have identical `repr` outputs for the two separate triggers.

`Event` and `Lock` have the address of each trigger in the `repr` output.

**Comment by ktbarrett** [comment]
> Both `<Task 1>` and `<Task 2>` wait on the same `Combine(Timer(1), Join(<Task 3>)` object.

I know our system allows you to do this, but I don't think it's ever a good idea. There are a lot of issues when it comes to exceptions if you do that, same thing if you await on the same task in more than one task. Future-like objects should only have one awaiting task, well defined programs form a tree.

**Comment by ktbarrett** [comment]
@garmin-mjames I'm happy, are you?

**Comment by ktbarrett** [comment]
Well, if you want to change something we still have several months before the release.

**Comment by garmin-mjames** [comment]
@ktbarrett Yeah, I'm happy with where this is at now, unless you think `RunningTask.__repr__` needs a docstring.
